Switch "representation" to "representation_type" in coordinates

adrn opened the pull request:
The context is: the frame classes and `SkyCoord` accept a `representation=` keyword and have a `.representation` attribute (and similar for `differential`), but these are the *class*, not the underlying representation object (that is `.data`). This has confused me, @eteq, and at least a few users, so we have a long term plan of deprecating and removing these attributes in favor of `representation_type` and `differential_type`. 

This is something we discussed in the lead up to v2.0 and then recently at the coordination meeting. 

This still needs:
* [x] Documentation updates
* [x] Updates to the tests: right now I left the tests (for `representation` at least) just to make sure that they still pass with the new attribute name / keyword.
* [x] At least one check that this doesn't break things in the wild (@cadair checking this with Sunpy would be great!)
* [x] TODO comments over each line that will be deprecated in the future

This fixes #6591

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -236,17 +236,17 @@ class BaseCoordinateFrame(ShapedLikeNDArray, metaclass=FrameMeta):
     ----------
-    representation : `BaseRepresentation` or None
-        A representation object or `None` to have no data (or use the other
+    representation_type : `BaseRepresentation` or None
```
I think this one should be changed to `data` - since here you are allowed to put in a `BaseRepresentation` *instance*.

Although I always found it annoying that in fact the initializer is `*args`. But that is independent of this PR?

adrn replied:
Yea, the `*args* thing annoys me too...I wonder if there is any way to use the metaclass to hack the signature? Sounds a little scary though. But yea, I think outside of the scope of this PR.

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -236,17 +236,17 @@ class BaseCoordinateFrame(ShapedLikeNDArray, metaclass=FrameMeta):
-        Coordinates, with names that depend on the subclass.
-    differential_cls : `BaseDifferential`, dict, optional
+        Coordinate components, with names that depend on the subclass.
+    differential_type : `BaseDifferential`, dict, optional
```
And above you need to add `representation_type` (as in the initializer below).

I also note that the line is confusing: it should be 
```
differential_type: `BaseDifferential` subclass, str, dict, optional
```

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -265,28 +265,34 @@ class BaseCoordinateFrame(ShapedLikeNDArray, metaclass=FrameMeta):
+        # TODO: this is here for backwards compatibility. It should be possible
+        # to use either the kwarg representation_type, or representation.
+        # In future versions, we will raise a deprecation warning here:
+        representation_type = kwargs.pop('representation', representation_type)
```
This should be an error if `representation_type` is given as well. So,
```
if 'representation' in kwargs:
    if representation_type is not None:
        raise
    representation_type  = kwargs.pop('representation')
```

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -265,28 +265,34 @@ class BaseCoordinateFrame(ShapedLikeNDArray, metaclass=FrameMeta):
 
         # TODO: we should be able to deal with an instance, not just a
-        # class or string for representation and differential_cls.
+        # class or string for representation and differential_type.
```
This comment is actually misleading, as currently the class is able to deal with an instance for representation (i.e., the first argument can be a `BaseRepresentation` instance. So, the question would be whether we should be able to deal with a `Differential` instance as well. Personally, I don't think this is necessary - one can just pass in a representation with differential already attached.

adrn replied:
Ok! Removing this comment.

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -312,8 +318,8 @@ def __init__(self, *args, copy=True, representation=None,
-        elif self.representation:
-            representation_cls = self.representation
+        elif self.representation_type:
+            representation_type = self.representation_type
```
I think here it is actually clearer to stick with `representation_cls = self.get_representation_cls()` (i.e., use the getter), because the whole stanza below deals with something known to be a class.

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -330,16 +336,17 @@ def __init__(self, *args, copy=True, representation=None,
                     del repr_kwargs['distance']
 
-                if (issubclass(representation_cls, r.SphericalRepresentation)
+                if (issubclass(representation_type, r.SphericalRepresentation)
```
So, I'd undo this change and the others below.

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -330,16 +336,17 @@ def __init__(self, *args, copy=True, representation=None,
             # Get any differential data passed in to the frame initializer
             # using keyword or positional arguments for the component names
-            differential_cls = self.get_representation_cls('s')
+            differential_type = self.get_representation_cls('s')
```
Same here, just stick with `_cls`

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -783,27 +803,27 @@ def represent_as(self, base, s='base', in_frame_units=False):
         # We only keep velocity information
         if 's' in self.data.differentials:
-            differential_cls = repr_classes['s']
+            differential_type = repr_classes['s']
```
Here again I think `_cls` was appropriate, so I'd keep it here and below.

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -854,7 +874,7 @@ def represent_as(self, base, s='base', in_frame_units=False):
                     # Here we have to bypass using with_differentials() because
                     # it has a validation check. But because .representation and
-                    # .differential_cls don't point to the original classes, if
+                    # .differential_type don't point to the original classes, if
```
Last one of this part where it should remain `_cls`

mhvk commented on astropy/coordinates/builtin_frames/altaz.py:
```diff
@@ -51,12 +51,12 @@ class AltAz(BaseCoordinateFrame):
         Can only be passed in as a keyword argument.
 
-    differential_cls : `BaseDifferential`, dict, optional
+    differential_type : `BaseDifferential`, dict, optional
```
Strange that the `str` option was missing again. See comment for `Baseframe` in that it also should be `BaseDifferential` subclass. I guess this holds for all other frames too. Sigh.

It may be worth using `astropy.utils.decorators.format_doc`...

adrn replied:
OK, I sucked it up and overhauled all of the docstrings to use `format_doc`. I did it in a single commit, so it's probably easiest to look at the diff for that commit...

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -217,7 +219,11 @@ def __init__(self, *args, copy=True, **kwargs):
         coord_kwargs = {}
         if 'representation' in kwargs:
-            coord_kwargs['representation'] = _get_repr_cls(kwargs['representation'])
+            kwargs['representation_type'] = kwargs.pop('representation')
```
I'd raise if `representation_type` is present already in `kwargs`

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -320,8 +336,14 @@ def _parse_inputs(self, args, kwargs):
+        # TODO: deprecate this in future
         if 'representation' in kwargs:
-            valid_kwargs['representation'] = _get_repr_cls(kwargs.pop('representation'))
+            valid_kwargs['representation_type'] = _get_repr_cls(
```
Again, raise if `representation_type` is already present.

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1594,8 +1616,15 @@ class like ICRS, but not an instance ICRS() since the latter could have
+    # TODO: deprecate this in future
     if 'representation' in kwargs:
-        frame = frame_cls(representation=_get_repr_cls(kwargs['representation']))
+        repr_cls = _get_repr_cls(kwargs['representation'])
```
And again. Also, I'd set `representation_type` as you have above. It does seem like you could make a mini-helper function `_normalize_representation_type(kwargs)`...

mhvk commented on astropy/coordinates/tests/test_frames_with_velocity.py:
```diff
@@ -176,12 +176,12 @@ def test_differential_cls_arg():
     icrs = ICRS(ra=1*u.deg, dec=60*u.deg,
                 pm_ra=10*u.mas/u.yr, pm_dec=-11*u.mas/u.yr,
-                differential_cls=r.UnitSphericalDifferential)
+                differential_type=r.UnitSphericalDifferential)
```
Hmm, don't we test passing in a string?

adrn replied:
I added a test for this.

Cadair commented:
I do need to check this works with SunPy. The change will mean some refactoring for us when we are astropy >= 3.0 only. (Which will not be our next release as we have one more 2.7 release scheduled)

adrn commented:
@Cadair This PR shouldn't break sunpy (we should test this when I get farther along) as it doesn't deprecate or remove using "`representation`", it just adds the new name `representation_type` as the name that actually stores the info. Setting/getting `.representation` will just pass through to `representation_type`. This is just the first step towards eventually deprecating and removing.

adrn commented:
OK I think I addressed @mhvk's comments. The diffs are pretty ugly - sorry about that...One major thing I did (in [653395e](https://github.com/astropy/astropy/pull/6873/commits/653395ec0cc7766adad87ad889a87fb37c98fb0f)) is to overhaul the builtin frame docstrings to use the `format_doc()` decorator. 

I added a test for backwards compatibility, and improved one of the frame initialization tests that tries a bunch of valid argument combinations. Next commit(s) will be documentation updates.

taldcroft commented:
From a brief scan of diffs from the API perspective this looks good to me and matches how we discussed handling this at the coordination meeting.  Thanks!

@Cadair - your SunPy testing will be crucial as the idea we agreed on is that this change should not break any code right now, even testing that is being strict about warnings.  (I.e. in this release there won't even be pending deprecation warnings, just doc / example updates that reflect the new preferred API).

@mhvk - yes, the idea is to take this very slowly.

taldcroft commented:
This change should be explicitly discussed somewhere in the narrative docs.  Maybe a `.. note::` at the top of the Reference/API section in the coordinates `index.rst`?  We don't always bump things like this up to the main docs, but in this case I can imagine people just being confused that this keyword has changed without explanation.

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -90,6 +91,20 @@ def _represent_as_dict(self):
         return out
 
 
+def _normalize_representation_type(kwargs):
```
Move this to baseframe, use it there as well, and import it here? Would help ensure that we deprecate at the same time.

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -234,13 +254,22 @@ def __init__(self, *args, copy=True, **kwargs):
+    # TODO: deprecate these in future
     @property
     def representation(self):
-        return self.frame.representation
```
Might leave these as is, again to ensure we do things at the same time.

mhvk commented on astropy/coordinates/tests/test_frames_with_velocity.py:
```diff
@@ -60,6 +73,8 @@ def test_all_arg_options(kwargs):
     repr_gal = repr(gal)
 
     for k in kwargs:
+        if '_type' in k:
```
Why? Shouldn't only `differential_type` be skipped?

eteq commented on astropy/coordinates/baseframe.py:
```diff
@@ -265,28 +292,36 @@ class BaseCoordinateFrame(ShapedLikeNDArray, metaclass=FrameMeta):
+        # This is here for backwards compatibility. It should be possible
+        # to use either the kwarg representation_type, or representation.
+        # TODO: In future versions, we will raise a deprecation warning here:
+        _normalize_representation_type(kwargs)
```
this looks to me to be the source of all the failures: `_normalize_representation_type` is expecting `representation_type` to be in the `kwargs`, but it isn't - it's a separate stand-alone keyword.  If I add this immediately above:
```
        if 'representation' in kwargs:
            representation_type = kwargs.pop('representation')
```
most of the tests begin to pass.  So from that I derived the commit I pushed up to this branch that I hope should get the tests all passing.

adrn commented:
Why does this need a changelog entry in new features? I added a line in the API changes section. Can you add one to new features if you think it is necessary?

Your commit seems to have fixed the build failures - thanks

adrn commented:
Oof, that was a pretty nasty rebase. I think I cleaned up any possible issues, but it would be good for someone to take a look at the line changes again. The main files with conflicts were `baseframe.py` and `sky_coordinate.py`.

I also added a `.. note::` to the documentation explaining the change.

eteq commented:
(I guess I didn't say this above, but I looked and other than a changelog tweak it looked good to me - @adrn seems to have strong rebase-fu!)
